lib/spack/llnl/util/symlink.py

```diff
@@ -2,77 +2,166 @@
 
 if sys.platform == "win32":
     from win32file import CreateHardLink
 
 
-def symlink(real_path, link_path):
+def symlink(source_path: str, link_path: str):
     """
-    Create a symbolic link.
+    Try to create a symbolic link.
+
+    On non-Windows and Windows with System Administrator
+    privleges this will be a normal symbolic link via
+    os.symlink.
+
+    On Windows without privledges the link will be a
+    junction for a directory and a hardlink for a file.
+    On Windows the various link types are:
+
+    Symbolic Link: A link to a file or directory on the
+    same or different volume (drive letter) or even to
+    a remote file or directory (using UNC in its path).
+    Need System Administrator privileges to make these.
 
-    On Windows, use junctions if os.symlink fails.
+    Hard Link: A link to a file on the same volume (drive
+    letter) only. Every file (file's data) has at least 1
+    hard link (file's name). But when this method creates
+    a new hard link there will be 2. Deleting all hard
+    links effectively deletes the file. Don't need System
+    Administrator privileges.
+
+    Junction: A link to a directory on the same or different
+    volume (drive letter) but not to a remote directory. Don't
+    need System Administrator privileges.
     """
-    if sys.platform != "win32":
-        os.symlink(real_path, link_path)
-    elif _win32_can_symlink():
-        # Windows requires target_is_directory=True when the target is a dir.
-        os.symlink(real_path, link_path, target_is_directory=os.path.isdir(real_path))
+    source_path = os.path.normpath(source_path)
+    link_path = os.path.normpath(link_path)
+
+    # Perform basic checks to make sure symlinking will succeed
+    if os.path.exists(link_path):
+        raise SymlinkError(f"Link path ({link_path}) already exists. Cannot create link.")
+
+    if not os.path.exists(source_path):
+        if os.path.isabs(source_path):
+            # An absolute source path that does not exist will result in a broken link.
+            raise SymlinkError(
+                f"Source path ({source_path}) is absolute but does not exist. Resulting "
+                f"link would be broken so not making link."
+            )
+        else:
+            # os.symlink can create a link when the given source path is relative to
+            # the link path. Emulate this behavior and check to see if the source exists
+            # relative to the link patg ahead of link creation to prevent broken
+            # links from being made.
+            link_parent_dir = os.path.dirname(link_path)
+            relative_path = os.path.join(link_parent_dir, source_path)
+            if not os.path.exists(relative_path):
+                raise SymlinkError(
+                    f"The source path ({source_path}) is not relative to the link path "
+                    f"({link_path}). Resulting link would be broken so not making link."
+                )
```

Comment:
(question) will the symlink work in this case on Windows without rewriting `source_path`? If so, is that tested? I wouldn't be surprised if we had to update `source_path`

Reply:
I added additional coverage to the unit tests, but `symlink.py::test_symlink_src_relative_to_link` and `symlink.py::test_symlink_src_not_relative_to_link` should cover this case. Both pass on my Windows dev machine.